server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java

```diff
@@ -79,9 +69,15 @@ public class KnnSearchBuilder implements Writeable, ToXContentFragment, Rewritea
 
     static {
         PARSER.declareString(constructorArg(), FIELD_FIELD);
-        PARSER.declareFloatArray(optionalConstructorArg(), QUERY_VECTOR_FIELD);
+        PARSER.declareField(
+            optionalConstructorArg(),
+            (p, c) -> parseQueryVector(p),
+            QUERY_VECTOR_FIELD,
+            ObjectParser.ValueType.OBJECT_ARRAY_STRING_OR_NUMBER
```

Comment:
While I like us using this value type, we should not allow a single number. 

Additionally, we should take advantage of this to parse directly to `byte[]` values and serialize them. This can remove a nice bit of heap overhead. Though, it will require backwards compatible transport seralization. Meaning, if we are serializing to an old node, transform to a `float[]` and write it, otherwise write the `byte[]`.

Reply:
> So even though we could be handing a byte vector field, this information becomes available at a later point and it isn't straightforward to fail fast (if provided a hex or non-conforming values for byte[] vectors for example).

Yeah, I understand. I think we should do this:

 - Parse to `byte[]`
 - Transform to `float[]` only on older nodes
 - Allow `byte[]` to be transformed to `float[]` if the encoding type isn't `float` down on the mapping.

This gives us consistent behavior with what we are doing now & the added benefit of reduced overhead of serializing `byte[]` values everywhere instead of `float[]`

Reply:
Was looking on how to best support this and tried out 2 approaches:
* Add a new DTO that would define a `byteArray` and a `floatArray` only one of which would be non-null. Then use this in all places up until we reach the point of actually creating the query, and have helper methods to convert `float`->`byte`, `byte`-> `float` , and a best-effort parsing (i.e. parse the array and iff **all** elements can be properly converted to `byte`, assume that it's a `byte` array). For bwc, always write `float` arrays, and when reading try to interpret whether it's a `float` or `byte`.
* Add a new explicit field `byteVector` on `KnnSearchBuilder` and `KnnVectorQueryBuilder` (which we already had and was removed as part of https://github.com/elastic/elasticsearch/pull/98916/). This is a "simpler" approach, albeit with some needed code duplication. Changes are much narrower as they would only affect the `hex` case, and we can select which array to pass in `KnnVectorQueryBuilder#doToQuery` based on the specified `elementType` (or which flavor of the vectors is not null). 

Personally I think that the 1st approach would give us more flexibility into how we're parsing vectors, and at the same time keep the logic of converting / transforming from one type to another in a single place. However, it is a more extensive code change that may fall outside of the scope of this PR. 

@benwtrent wdyt? Should I go ahead with any of the above or do you have any other suggestions?

Reply:
@pmpailis I don't mind a record class that knows how to parse the appropriate value and handles all the BWC internally. The code would be small and contained

Reply:
The `byteVector` was removed because it was never actually used. +1 on idea to add DTO/record that holds `byteArray` and `floatArray` one of each is null. We have many of these examples in Lucene.

Reply:
Added an initial approach for this in 135d8d0500f0b05f5933e2e0e215f3087a2bb364